plugin-commons/src/main/java/org/sonarsource/sonarlint/core/plugin/commons/loading/PluginInstancesLoader.java

```diff
@@ -85,31 +95,44 @@ public Map<String, Plugin> instantiatePluginClasses(Collection<PluginInfo> plugi
   Collection<PluginClassLoaderDef> defineClassloaders(Map<String, PluginInfo> pluginsByKey) {
     Map<String, PluginClassLoaderDef> classloadersByBasePlugin = new HashMap<>();
 
-    for (PluginInfo info : pluginsByKey.values()) {
+    for (var info : pluginsByKey.values()) {
       var baseKey = basePluginKey(info, pluginsByKey);
       if (baseKey == null) {
         continue;
       }
       var def = classloadersByBasePlugin.computeIfAbsent(baseKey, PluginClassLoaderDef::new);
       def.addFiles(List.of(info.getJarFile()));
+      getJarFile(info.getJarFile().toPath()).ifPresent(jarFilesToClose::add);
       if (!info.getDependencies().isEmpty()) {
         LOG.warn("Plugin '{}' embeds dependencies. This will be deprecated soon. Plugin should be updated.", info.getKey());
         var tmpFolderForDeps = createTmpFolderForPluginDeps(info);
-        for (String dependency : info.getDependencies()) {
+        for (var dependency : info.getDependencies()) {
           var tmpDepFile = extractDependencyInTempFolder(info, dependency, tmpFolderForDeps);
           def.addFiles(List.of(tmpDepFile.toFile()));
           filesToDelete.add(tmpDepFile);
+          getJarFile(tmpDepFile).ifPresent(jarFilesToClose::add);
         }
       }
       def.addMainClass(info.getKey(), info.getMainClass());
 
-      for (String defaultSharedResource : DEFAULT_SHARED_RESOURCES) {
+      for (var defaultSharedResource : DEFAULT_SHARED_RESOURCES) {
         def.getExportMask().addInclusion(String.format("%s/%s/api/", defaultSharedResource, info.getKey()));
       }
     }
     return classloadersByBasePlugin.values();
   }
 
+  private static Optional<JarFile> getJarFile(Path tmpDepFile) {
+    try {
+      return Optional.of(((JarURLConnection) new URL("jar:" + tmpDepFile.toUri().toURL() + "!/").openConnection()).getJarFile());
+    } catch (ZipException ignore) {
+      // For tests, we are using fake JARs, so ignore ZipException: zip file is empty
+      return Optional.empty();
+    } catch (IOException e) {
+      throw new IllegalArgumentException(e);
```

Comment:
I think this would fail the loading process for all plugins, shouldn't we just ignore the plugin but go on with the rest?

Reply:
Hum, that was not the previous behavior AFAIK. What if a plugin depends on another plugin. We might end up in an inconsistent state.

Reply:
I might be wrong but it's a new method, so this is a change in behavior

Reply:
My point is that all Jar paths are supposed to exist. If they don't, it would have failed later when building the URLClassloader. So failing here or later is the same.

Reply:
The reason for my comment was [this line](https://github.com/SonarSource/sonarlint-core/blob/master/plugin-commons/src/main/java/org/sonarsource/sonarlint/core/plugin/commons/loading/PluginClassloaderFactory.java#L100), I have the impression we are only logging for ClassLoaders that cannot be built. This is something I implemented a while ago but it's maybe not the right way to do that

Reply:
I have a similar idea that all classloader definitions could depend on each other. Having a classloader failing to be built, and still trying to build/load others may create unexpected errors.
In the long run, if we feel the need to support the case of one analyzer failing to load, while still being able to run others, I think we should go in a direction where we totally isolate each analyzer into its own container (or even process?).